sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1066,12 +1075,144 @@ async Task performProcessing()
+
+            try
+            {
+                var connection = CreateConnection();
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var cycleDuration = new Stopwatch();
+                var partitionIds = default(string[]);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    cycleDuration.Restart();
+
+                    try
+                    {
+                        partitionIds = await connection.GetPartitionIdsAsync(retryPolicy, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (ex.IsNotType<TaskCanceledException>())
+                    {
+                        // Logging for exceptions with the service operation are responsibility of the connection.
+
+                        _ = InvokeOnProcessingErrorAsync(ex, null, Resources.OperationGetPartitionIds, CancellationToken.None);
+                        partitionIds = default;
+                    }
+
+                    // Perform a load balancing cycle only in the case that partitions were successfully read.
+
+                    if (partitionIds != default)
+                    {
+                        var claimedOwnership = default(EventProcessorPartitionOwnership);
+
+                        try
+                        {
+                            claimedOwnership = await LoadBalancer.RunLoadBalancingAsync(partitionIds, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (EventHubsException ex)
+                            when (Resources.OperationClaimOwnership.Equals(ex.GetFailureOperation(), StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            // If a specific partition was associated with the failure, it is carried by a well-known data member of the exception.
+
+                            var partitionId = ex.GetFailureData<string>();
+
+                            var partition = (partitionId ?? string.Empty) switch
+                            {
+                                string id when (id == string.Empty) => null,
+                                string _ when (ActivePartitionProcessors.TryGetValue(partitionId, out var partitionProcessor)) => partitionProcessor.Partition,
+                                _ => new TPartition { PartitionId = partitionId }
+                            };
+
+                            _ = InvokeOnProcessingErrorAsync(ex.InnerException ?? ex, partition, Resources.OperationClaimOwnership, CancellationToken.None);
+                            Logger.EventProcessorClaimOwnershipError(Identifier, EventHubName, ConsumerGroup, partitionId, ((ex.InnerException ?? ex).Message));
+                        }
+                        catch (Exception ex) when (ex.IsNotType<CancellationToken>())
```

Comment:
I guess this would have been impossible if T had a constraint on it being an exception type (which is what the documentation for the helper implied).  Maybe add that constraint to prevent stuff like this?

Reply:
gah.   That's... not awesome.  I had started with that extension constrained to only `Exception` types, but then opened it to all the things, as I can see it being useful in other contexts, albeit marginally.

Reply:
I'm going to leave it as-is for now, but I took a note to give it a look during the next round of changes and consider scoping it back to only exceptions.

Reply:
SGTM

Reply:
I lied; it ended up annoying me enough that I made the change now.